### ISSUE TITLE
Blocks execution if rabbitmq is not reachable

### ISSUE THREAD

**Issue by AntonioAlfrz** [nttcom/fluent-plugin-rabbitmq#11]
# Inquiry
If RabbitMQ host is not reachable, it blocks td-agent. Is this the intended behavior?
For example, in the following case, it will not print the dummy events.

# Configuration
```
<source>
  @type dummy
  tag mytest
  dummy {"hello":"world"}
</source>

<match **>
  @type stdout
</match>

<match **>
  @type rabbitmq
  host 192.168.0.1
  # or hosts ["192.168.1.1", "192.168.1.2"]
  user user
  pass password
  vhost /
  format json # or msgpack, ltsv, none
  exchange test # required: name of exchange
  exchange_type topic # required: type of exchange e.g. topic, direct
  exchange_durable false
  routing_key mytest # if not specified, the tag is used
  heartbeat 10 # integer as seconds or :server (interval specified by server)
</match>
```
# Plugins
```
bunny (2.13.0)
fluent-plugin-rabbitmq (0.0.5)
fluentd (1.3.3)

```

# Output

```
2019-01-25 07:27:44 +0000 [warn]: #0 fluent/log.rb:342:warn: Could not establish TCP connection to 192.168.0.1:5672: Bunny::ClientTime out                                                                                                                                    2019-01-25 07:27:44 +0000 [error]: #0 fluent/log.rb:362:error: unexpected error error_class=Bunny::TCPConnectionFailedForAllHosts erro r="Could not establish TCP connection to any of the configured hosts"                                                                    2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ bunny-2.13.0/lib/bunny/session.rb:333:in `rescue in start'                                                                               2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ bunny-2.13.0/lib/bunny/session.rb:305:in `start'                                                                                         2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluent-plugin-rabbitmq-0.0.5/lib/fluent/plugin/out_rabbitmq.rb:119:in `start'                                                            2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/root_agent.rb:165:in `block in start'                                                                           2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/root_agent.rb:154:in `block (2 levels) in lifecycle'                                                            2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/root_agent.rb:153:in `each'                                                                                     2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/root_agent.rb:153:in `block in lifecycle'                                                                       2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/root_agent.rb:140:in `each'                                                                                     2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/root_agent.rb:140:in `lifecycle'                                                                                2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/root_agent.rb:164:in `start'                                                                                    2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/engine.rb:274:in `start'                                                                                        2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/engine.rb:219:in `run'                                                                                          2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/supervisor.rb:799:in `run_engine'                                                                               2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/supervisor.rb:549:in `block in run_worker'                                                                      2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/supervisor.rb:724:in `main_process'                                                                             2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/supervisor.rb:544:in `run_worker'                                                                               2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/lib/fluent/command/fluentd.rb:316:in `'                                                                    2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/ rubygems/core_ext/kernel_require.rb:55:in `require'                                                                                      2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/ rubygems/core_ext/kernel_require.rb:55:in `require'                                                                                      2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/ fluentd-1.3.3/bin/fluentd:8:in `'                                                                                        2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/bin/fluentd:23:in `load'    2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /opt/td-agent/embedded/bin/fluentd:23:in ` required)>'                                                                                                                              2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /sbin/td-agent:7:in `load'                         2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:549:block in run_worker: /sbin/td-agent:7:in `'                     2019-01-25 07:27:44 +0000 [error]: #0 fluent/log.rb:362:error: unexpected error error_class=Bunny::TCPConnectionFailedForAllHosts erro r="Could not establish TCP connection to any of the configured hosts"                                                                    2019-01-25 07:27:44 +0000 [error]: #0 fluent/supervisor.rb:723:main_process: suppressed same stacktrace                              2019-01-25 07:27:44 +0000 [info]: fluent/log.rb:322:info: Worker 0 finished unexpectedly with status 1
```

**Comment by mmasaki** [comment]
As shown in the output, the error was raised in `start` method in the plugin.
I think this is an intended behavior of fluentd (td-agent).

**Comment by AntonioAlfrz** [comment]
Alright, thank you for your response.
Does it have buffer features in case rabbitmq is not reachable?

**Comment by mmasaki** [comment]
No. It is an non-buffered plugin.
However, I'm considering to support buffer and async buffer features.

**Comment by AntonioAlfrz** [comment]
Thank you, I am looking forward to it.

**Comment by mmasaki** [comment]
Hi. Just for your notice, I released v0.0.6 and it supports buffered output. Enjoy!

**Comment by AntonioAlfrz** [comment]
Great! Thank you!
